metricbeat/scripts/mage/docs_collector.go

```diff
@@ -0,0 +1,413 @@
+		if err != nil {
+			return nil, err
+		}
+
+		// generate the asciidoc link used in the module docs, since we need this in a few places
+		link := fmt.Sprintf("<<metricbeat-metricset-%s-%s,%s>>", moduleName, metricsetName, metricsetName)
+
+		// test to see if the metricset has a data.json
+		hasData := false
+		_, err = os.Stat(filepath.Join(metricset, "_meta/data.json"))
+		if err == nil {
+			hasData = true
+		}
+
+		ms := metricsetData{
+			Path:       metricset,
+			Title:      metricsetName,
+			Release:    release,
+			Link:       link,
+			DataExists: hasData,
+		}
+
+		metricsets = append(metricsets, ms)
+
+	} // end of metricset loop
+
+	return metricsets, nil
+}
+
+// gatherData gathers all the data we need to construct the docs that end up in metricbeat/docs
+func gatherData(modules []string) ([]moduleData, error) {
+	moduleList := make([]moduleData, 0)
+	//iterate over all the modules, checking to make sure we have an asciidoc file
+	for _, module := range modules {
+
+		isModule, err := testIfDocsInDir(module)
+		if err != nil {
+			return moduleList, err
+		}
+		if !isModule {
+			continue
+		}
+		moduleName := filepath.Base(module)
+
+		err = createDocsPath(moduleName)
+		if err != nil {
+			return moduleList, err
+		}
+
+		fieldsm, err := loadModuleFields(filepath.Join(module, "_meta/fields.yml"))
+		if err != nil {
+			return moduleList, err
+		}
+
+		cfgPath, err := getConfigfile(module)
+		if err != nil {
+			return moduleList, err
+		}
+
+		metricsets, err := gatherMetricsets(module, moduleName)
+		if err != nil {
+			return moduleList, err
+		}
+
+		//dump the contents of the module asciidoc
+		moduleDoc, err := ioutil.ReadFile(filepath.Join(module, "_meta/docs.asciidoc"))
+		if err != nil {
+			return moduleList, err
+		}
+
+		fieldsm.Path = module
+		fieldsm.CfgFile = cfgPath
+		fieldsm.Metricsets = metricsets
+		fieldsm.Asciidoc = string(moduleDoc)
+		fieldsm.Dashboards = hasDashboards(module)
+		fieldsm.Base = moduleName
+
+		moduleList = append(moduleList, fieldsm)
+
+	} // end of modules loop
+
+	return moduleList, nil
+}
+
+// writeModuleDocs writes the module-level docs
+func writeModuleDocs(modules []moduleData, t *template.Template) error {
+	for _, mod := range modules {
+		filename := mage.OSSBeatDir(filepath.Join("docs", "modules", fmt.Sprintf("%s.asciidoc", mod.Base)))
+		err := writeTemplate(filename, t.Lookup("moduleDoc.tmpl"), mod)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// writeMetricsetDocs writes the metricset-level docs
+func writeMetricsetDocs(modules []moduleData, t *template.Template) error {
+	for _, mod := range modules {
+		for _, metricset := range mod.Metricsets {
+			modData := struct {
+				Mod       moduleData
+				Metricset metricsetData
+			}{
+				mod,
+				metricset,
+			}
+			filename := mage.OSSBeatDir(filepath.Join("docs", "modules", mod.Base, fmt.Sprintf("%s.asciidoc", metricset.Title)))
```

Comment:
`OSSBeatDir` joins paths passed as parameters, so this should be equivalent:
```suggestion
			filename := mage.OSSBeatDir("docs", "modules", mod.Base, fmt.Sprintf("%s.asciidoc", metricset.Title))
```
Same in other cases.